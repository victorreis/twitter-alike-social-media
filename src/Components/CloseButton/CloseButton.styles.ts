import styled from 'styled-components';

import { CloseButtonStyleProps } from './CloseButton.types';

// The icon library does not pass the onClick to the SVG component,
// so I need to create this transparent container.
export const TransparentContainer = styled.div`
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 1;
`;

export const CloseButtonContainer = styled.button<CloseButtonStyleProps>`
  cursor: pointer;
  background-color: transparent;
  width: 2.5rem;
  height: 2.5rem;
  margin-top: -1.5rem;
  margin-right: -1rem;
  align-self: flex-end;
  border: 0;
  filter: drop-shadow(
      1px 1px 1px ${({ theme }) => theme.colors.main.effect.darker}
    )
    drop-shadow(-2px -2px 2px ${({ theme }) => theme.colors.main.effect.darker});
`;
