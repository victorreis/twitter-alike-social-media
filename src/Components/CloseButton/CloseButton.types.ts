import { TestProps } from '../../Config/Tests/Test.types';

export interface RequiredCloseButtonProps {
  /**
   * Callback function that is called when the component is clicked.
   */
  onClick: () => void;
}

export interface DefaultCloseButtonProps {}

export interface OptionalCloseButtonProps {}

export type CloseButtonProps = RequiredCloseButtonProps &
  DefaultCloseButtonProps &
  OptionalCloseButtonProps &
  TestProps &
  Omit<React.HTMLAttributes<HTMLButtonElement>, 'children' | 'onClick'>;

export type CloseButtonStyleProps = Required<DefaultCloseButtonProps>;
