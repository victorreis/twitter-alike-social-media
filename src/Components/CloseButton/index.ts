export * from './CloseButton.types';
export * from './CloseButton';
