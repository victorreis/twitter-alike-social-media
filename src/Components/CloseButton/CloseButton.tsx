import { MdOutlineClose } from 'react-icons/md';

import { useTheme } from 'styled-components';

import { TestProps } from '../../Config/Tests/Test.types';
import { iconSizes } from '../../Theme/Types';
import {
  CloseButtonContainer,
  TransparentContainer,
} from './CloseButton.styles';
import { CloseButtonProps, DefaultCloseButtonProps } from './CloseButton.types';

export const closeButtonDefaults: Required<DefaultCloseButtonProps> &
  Required<TestProps> = {
  testID: 'CloseButton',
};

export const CloseButton: React.FC<CloseButtonProps> = (props): JSX.Element => {
  const theme = useTheme();
  const { testID = closeButtonDefaults.testID, onClick, ...others } = props;

  return (
    <CloseButtonContainer data-testid={testID} onClick={onClick} {...others}>
      <MdOutlineClose
        color={theme.colors.background.default.lightest}
        onClick={onClick}
        size={iconSizes.md}
      />
      <TransparentContainer />
    </CloseButtonContainer>
  );
};
