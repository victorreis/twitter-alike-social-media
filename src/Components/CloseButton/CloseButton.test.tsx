import {
  fireEvent,
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { CloseButton, closeButtonDefaults } from './CloseButton';
import {
  RequiredCloseButtonProps,
  CloseButtonProps,
} from './CloseButton.types';

describe('closeButton component tests', () => {
  const onClick = jest.fn();

  const requiredProps: RequiredCloseButtonProps = {
    onClick,
  };

  const setup = (props?: CloseButtonProps) => {
    const renderRTR = () =>
      renderRTRCreator<CloseButtonProps>(CloseButton, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<CloseButtonProps>(CloseButton, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it('should render the component', () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(closeButtonDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it('should call the onCLick callback when the button is clicked', () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(closeButtonDefaults.testID);

      fireEvent.click(testInstance);

      expect(onClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('snapshot tests', () => {
    it('should render correctly', () => {
      expect.assertions(1);

      const generatedJson = setup().renderRTR().toJSON();
      expect(generatedJson).toMatchSnapshot();
    });
  });
});
