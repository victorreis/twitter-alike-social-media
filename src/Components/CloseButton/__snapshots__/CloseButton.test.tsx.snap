// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`closeButton component tests snapshot tests should render correctly 1`] = `
<button
  className="CloseButtonContainer-twitter-alike-social-media__sc-1gjf1sx-1 fzoSSR"
  data-testid="CloseButton"
  onClick={[MockFunction]}
>
  <svg
    color="#e8e8ee"
    fill="currentColor"
    height="1.67rem"
    onClick={[MockFunction]}
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": "#e8e8ee",
      }
    }
    viewBox="0 0 24 24"
    width="1.67rem"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0V0z"
      fill="none"
    />
    <path
      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
    />
  </svg>
  <div
    className="TransparentContainer-twitter-alike-social-media__sc-1gjf1sx-0 eysuIy"
  />
</button>
`;
